[SPARK-11053] Remove use of KVIterator in SortBasedAggregationIterator

JoshRosen opened the pull request:
SortBasedAggregationIterator uses a KVIterator interface in order to process input rows as key-value pairs, but this use of KVIterator is unnecessary, slightly complicates the code, and might hurt performance. This patch refactors this code to remove the use of this extra layer of iterator wrapping and simplifies other parts of the code in the process.

JoshRosen committed:
Simplify wrapping of iterator into KVIterator.

JoshRosen committed:
Remove KVIterator from SortBasedAggregationIterator

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43535 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43535/consoleFull) for   PR 9066 at commit [`2b8bf94`](https://github.com/apache/spark/commit/2b8bf947b0434146131b3af4a314cc89ace8a874).

SparkQA commented:
[Test build #43535 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43535/console) for   PR 9066 at commit [`2b8bf94`](https://github.com/apache/spark/commit/2b8bf947b0434146131b3af4a314cc89ace8a874).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43535/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

JoshRosen commented:
/cc @yhuai @rxin for review.

rxin commented:
Looks ok to me. I will leave it to @yhuai for the detailed review.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43538 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43538/consoleFull) for   PR 9066 at commit [`2b8bf94`](https://github.com/apache/spark/commit/2b8bf947b0434146131b3af4a314cc89ace8a874).

SparkQA commented:
[Test build #43538 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43538/console) for   PR 9066 at commit [`2b8bf94`](https://github.com/apache/spark/commit/2b8bf947b0434146131b3af4a314cc89ace8a874).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43538/
Test FAILed.

JoshRosen commented:
jenkins retest this please

AmplabJenkins commented:
Merged build triggered.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortBasedAggregationIterator.scala:
```diff
@@ -101,18 +117,16 @@ class SortBasedAggregationIterator(
-      val currentRow = inputKVIterator.getValue
+      val inputRow = inputIterator.next()
+      val groupingKey = groupingKeyProjection(inputRow).copy()
+      val currentRow = inputRow.copy()
```
Maybe we can avoid of these two `copy`s?

JoshRosen replied:
Good catch. We only need these copies in `initialize` since we're buffering the result there. I'll fix this shortly.

yhuai commented:
Looks good.

JoshRosen committed:
Remove unnecessary copy()

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43551 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43551/consoleFull) for   PR 9066 at commit [`ef53862`](https://github.com/apache/spark/commit/ef538621f7415a8502615181f6fb08721a809ff6).

SparkQA commented:
[Test build #43548 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43548/consoleFull) for   PR 9066 at commit [`2b8bf94`](https://github.com/apache/spark/commit/2b8bf947b0434146131b3af4a314cc89ace8a874).

SparkQA commented:
[Test build #43551 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43551/console) for   PR 9066 at commit [`ef53862`](https://github.com/apache/spark/commit/ef538621f7415a8502615181f6fb08721a809ff6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43551/
Test PASSed.

SparkQA commented:
[Test build #43548 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43548/console) for   PR 9066 at commit [`2b8bf94`](https://github.com/apache/spark/commit/2b8bf947b0434146131b3af4a314cc89ace8a874).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43548/
Test PASSed.

JoshRosen commented:
@yhuai thanks for reviewing. I've addressed your comment by removing the unnecessary `copy()` calls and the tests are still passing, so I'm going to merge this now.
